openedx/core/djangoapps/discussions/models.py

```diff
@@ -0,0 +1,53 @@
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from jsonfield import JSONField
+from opaque_keys.edx.django.models import LearningContextKeyField
+
+from openedx.core.djangoapps.config_model_utils.models import CourseAppConfigOptionsModel
+
+
+class DiscussionConfig(CourseAppConfigOptionsModel):
+    """
+    Configuration model to store configuration for Discussions applications.
+    """
+
+    provider = models.CharField(
+        blank=False,
+        db_index=True,
+        max_length=100,
+        verbose_name=_("Discussion provider"),
+        help_text=_("The discussion tool/provider."),
+    )
+    config = JSONField(
+        blank=False,
+        default={},
+        help_text=_("The discussion configuration data that can be user visible."),
+    )
+    private_config = JSONField(
```

Comment:
Can you provide some examples on how this will be used?

Will the private and public dictionaries be merged? Which takes precedence?
Will "child" configs (say, the full org, course, _and_ run are specified) "inherit" values from the "parent" (say, only org/site is specified) or will they just be used to "seed" the child at create-time (say, copy the settings from parent to child).

I see the general idea of what we're getting at w/ this hierarchy, but I think I'm still missing a few pieces...

Reply:
I created two separate public and private configuration objects since I wanted to segregate secret configuration such as API keys, client keys/secrets and general configuration settings, such as project ID, etc. This allows a admin will access to secrets to set these up, and for other users to modify non-secret configuration without needing access to or seeing the secret config.

As such these will not be merged, and plugins will get them as separate values so they can handle them separately. 

There is no inheritance of values, each edit produces a new config and the final value is used. Eventually you will be able to clone a config and make a new copy with a new unique ID, but that will break the link. 

I think many plugins won't need the non-secret config since course-related config will be stored elsewhere, and I can understand eliminating it for now and only adding it in the future if a need arises. 

Example use:

    For Discourse the configurations could be as follows:
        Secret Config: { 
            api_key: 6f2343a1-d7d5-4994-a86b-a693fd980937
        }
        Config: {
            base_url: https://discourse.somesite.com
            auth_mode: LTI/SSO
            sync_moderators: true
            course_admins_can_moderate: false
        }

Reply:
Cool, in that case, as long as we have the option to extend in the future as needed, I think we can punt on the separate config for now, to avoid the complication (both code + UI/UX)